### PR TITLE
rework locking in SslStream to support TLS1.3

### LINKED ISSUE #32920: [System.Net.Security] Potential bug in SslStream.Implementation.cs
Discovered while annotating the library #32541.

It looks like the `lazyResult` value will never get initialized in this [snippet](https://github.com/dotnet/runtime/blob/d76ef042f8ead9d06a447ab2b1004ae626185ca2/src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs#L429-L443). Similarly [here](https://github.com/dotnet/runtime/blob/d76ef042f8ead9d06a447ab2b1004ae626185ca2/src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs#L510-L524).

### PR DESCRIPTION
**Pull request by wfurt**
The locking code in SslStream was broken for quite a while without us noticing. We bump to it with OpenSSL 1.1.1 and TLS1.3 but the fix was to add locking to PAL: https://github.com/dotnet/corefx/pull/37736. I run to the same problems when working on #1720. When MessageDecrypt decides to start renegotiation (there is no real renegotiation in TLS1.3 but on Windows, we still do get  SEC_I_RENEGOTIATE as we need to pass data to `lsass` process. 

I started to fix it but after chatting with @stephentoub we decided to go with a simplistic approach: We grab synchronous lock around DecryptData/EncryptData and if renegotiation is happening we would use create TaskCompletionSource encrypt can wait on for completion. 

With this, the code is pretty simple and old craft can be removed so as the change to OpenSSL PAL. 

Performance seems to be about the same. On Windows, I get results bellow and it changes slightly between runs. On Linux, there was already lock and performance remains the same as well.

|                          Method |                                                                                                                  Toolchain |        Mean |     Error |    StdDev |      Median |         Min |         Max | Ratio | RatioSD |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|-------------------------------- |--------------------------------------------------------------------------------------------------------------------------- |------------:|----------:|----------:|------------:|------------:|------------:|------:|--------:|-------:|------:|------:|----------:|
|       DefaultHandshakeIPv4Async | master\CoreRun.exe |    780.6 us |  18.41 us |  21.20 us |    780.2 us |    750.4 us |    821.4 us |  1.00 |    0.03 | 3.1250 |     - |     - |  13.62 KB |
|       DefaultHandshakeIPv4Async |  newLocking\CoreRun.exe |    781.1 us |  17.16 us |  19.76 us |    777.3 us |    743.8 us |    811.4 us |  1.00 |    0.00 |      - |     - |     - |  13.68 KB |
|                                 |                                                                                                                            |             |           |           |             |             |             |       |         |        |       |       |           |
|       DefaultHandshakeIPv6Async | master\CoreRun.exe |    767.2 us |  14.24 us |  13.98 us |    770.5 us |    726.9 us |    783.2 us |  0.99 |    0.04 | 2.8409 |     - |     - |  13.61 KB |
|       DefaultHandshakeIPv6Async |  newLocking\CoreRun.exe |    769.2 us |  20.27 us |  23.34 us |    773.5 us |    733.6 us |    800.2 us |  1.00 |    0.00 | 2.9762 |     - |     - |  13.68 KB |
|                                 |                                                                                                                            |             |           |           |             |             |             |       |         |        |       |       |           |
|       DefaultHandshakePipeAsync | master\CoreRun.exe |    763.1 us |  18.29 us |  21.06 us |    760.5 us |    731.8 us |    804.2 us |  1.02 |    0.04 | 2.9762 |     - |     - |   15.8 KB |
|       DefaultHandshakePipeAsync |  newLocking\CoreRun.exe |    749.4 us |  16.23 us |  18.69 us |    742.2 us |    724.0 us |    783.1 us |  1.00 |    0.00 | 3.1250 |     - |     - |  15.86 KB |
|                                 |                                                                                                                            |             |           |           |             |             |             |       |         |        |       |       |           |
| TLS12HandshakeECDSA256CertAsync | master\CoreRun.exe |  4,279.7 us |  53.66 us |  50.19 us |  4,267.0 us |  4,210.0 us |  4,398.6 us |  1.00 |    0.02 |      - |     - |     - |   18.8 KB |
| TLS12HandshakeECDSA256CertAsync |  newLocking\CoreRun.exe |  4,289.8 us |  85.58 us |  80.06 us |  4,266.8 us |  4,187.0 us |  4,488.5 us |  1.00 |    0.00 |      - |     - |     - |  18.79 KB |
|                                 |                                                                                                                            |             |           |           |             |             |             |       |         |        |       |       |           |
|  TLS12HandshakeRSA1024CertAsync | master\CoreRun.exe |  3,713.6 us |  33.65 us |  29.83 us |  3,714.3 us |  3,635.7 us |  3,747.2 us |  1.00 |    0.02 |      - |     - |     - |  18.98 KB |
|  TLS12HandshakeRSA1024CertAsync |  newLocking\CoreRun.exe |  3,722.3 us |  68.32 us |  60.56 us |  3,723.6 us |  3,621.7 us |  3,827.1 us |  1.00 |    0.00 |      - |     - |     - |  19.11 KB |
|                                 |                                                                                                                            |             |           |           |             |             |             |       |         |        |       |       |           |
|  TLS12HandshakeRSA2048CertAsync | master\CoreRun.exe |  4,624.0 us |  45.38 us |  42.45 us |  4,636.7 us |  4,524.7 us |  4,697.6 us |  1.00 |    0.01 |      - |     - |     - |  19.37 KB |
|  TLS12HandshakeRSA2048CertAsync |  newLocking\CoreRun.exe |  4,604.8 us |  47.08 us |  44.03 us |  4,602.3 us |  4,542.0 us |  4,691.7 us |  1.00 |    0.00 |      - |     - |     - |  19.42 KB |
|                                 |                                                                                                                            |             |           |           |             |             |             |       |         |        |       |       |           |
|  TLS12HandshakeRSA4096CertAsync | master\CoreRun.exe | 10,789.4 us | 187.44 us | 175.33 us | 10,735.4 us | 10,540.6 us | 11,047.1 us |  0.97 |    0.02 |      - |     - |     - |  20.16 KB |
|  TLS12HandshakeRSA4096CertAsync |  newLocking\CoreRun.exe | 11,072.6 us | 186.93 us | 165.71 us | 11,052.7 us | 10,839.1 us | 11,502.9 us |  1.00 |    0.00 |      - |     - |     - |  20.17 KB |

I did run SslStress for a while and I did not see any issues. 

fixes #32920
contributes to #1720

### CODE REVIEW

File: src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs
Comment (stephentoub) [inline src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs]:
```suggestion
        private readonly object _handshakeLock = new object();
```

Are there any other objects SslStream always creates and doesn't share out that we could use for this purpose instead?

File: src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs
Comment (stephentoub) [inline src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs]:
```suggestion
        private TaskCompletionSource<bool>? _handshakeWaiter;
```

File: src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs
Comment (stephentoub) [inline src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs]:
```suggestion
```

File: src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs
Comment (stephentoub) [inline src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs]: What prevents this from stomping on a write to _handshakeWaiter from another thread?

File: src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs
Comment (stephentoub) [inline src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs]:
```suggestion
                    if (waiter != null)
```

File: src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs
Comment (stephentoub) [inline src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs]:
```suggestion
                    return _context!.Encrypt(buffer, ref outBuffer, out outSize);
```

File: src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs
Comment (wfurt) [reply src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs]: yes, I was thinking about using _context but in the end I decided to keep it separate for now. I can do more testing and I think it should be possible.

File: src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs
Comment (stephentoub) [inline src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs]: What happens when EncryptData is being called from an async call path... shouldn't we be awaiting this rather than blocking synchronously?

If for some reason that's not necessary, a TCS is overkill and it can just use a ManualResetEventSlim.  But it seems like a TCS is actually necessary because we want to enable synchronous code paths to block and asynchronous code paths to await.

File: src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs
Comment (stephentoub) [inline src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs]: I must be missing it... where is _handshakeWaiter completed? i.e. where is {Try}SetResult called?

File: src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs
Comment (stephentoub) [inline src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs]: I love that all of this gets deleted.

One step closer to removing LazyAsyncResult entirely :)

File: src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs
Comment (stephentoub) [reply src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs]: I'm fine if you want to go with a dedicated object for now.  We can revisit it when the dust settles a little.

File: src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs
Comment (stephentoub) [inline src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs]: This async method is no longer necessary, as it's just calling one method and awaiting it.  If you want to keep the method for some reason, remove "async" from its signature, and change the await to just be a return.  Otherwise, the callsites can just be changed to call directly the thing being awaited.

File: src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs
Comment (wfurt) [reply src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs]: I revoked the logic and added LockAsync to the adapter interface. On the "normal" path we should not hit this e.g. attempt to write/encrypt does not conflict with a renegotiation. 

If it does, we wait for the lock and we try it again. In Sync path, everything should be synchronous. For Async we go via async CompleteLockAndWriteAsync()

File: src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs
Comment (wfurt) [reply src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs]: It seems like I broke it in the final cleanup ;(

This is set in ReplyOnReAuthenticationAsync() now.

The sequence is:

1.  we try to decrypt under lock. 
2. if renegotiation is requested we create TaskCompletionSource still under lock. 
3. that gets cleaned up when a renegotiation is finished. 

4. On write side, we either win and there is no contention and waiting.
5. We get there when a renegotiation is pending. We wait (possibly in loop) until renegotiation is completed as there is no guarantee there is only one. However, in n practice, that is the case so we should be able to finish write/encrypt  as soon as TaskCompletionSource is triggered.
6. Thre are weird cases that renegotiation in TLS1.3 can finish without IO e.g. we get a frame to update session. That needs to be processed but it can finish very quickly. In that case, _handshakeWaiter should be null even if Encrypt failed with TryAgin under lock.

File: src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs
Comment (wfurt) [reply src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs]: see the comment above.

File: src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs
Comment (wfurt) [reply src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs]: I probably don't understand the comment. There is always only one read and one write at max. That is enforced via atomic _nestedRead and _nestedWrite.  Please take a look at updated processing.

File: src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs
Comment (stephentoub) [inline src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs]: This local async function is capturing data from the outer function.  I think if you look at the compiled IL, you'll find this is causing allocations even when the async local function is never invoked, e.g. note the display class allocation at the beginning of M here:
https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA+ABATARgLABQGADAAQY4B0AQgK4Bm9MUAzgNyEnk4CsHBnAMzkspAMKkA3oVKzywjElIBZABQBKKTLk7qATwAuMANoBdUsAZMopALykAglCgBDPQAUIEADYAeYIYwAHyUAMoAFi6wACaUAEowAHYGqjjq/Do6GADspABiXhoZmbLaJRgAHORKBRAaZZnSBCXlAJzVlACaAJYw3tFFDS2Ozm6ePr76RiERUTCxCQa0UImqlozM6UNyAL4NewQ7QA
You should pass in any state you need explicitly (capturing anything other than `this` is likely to cause such behavior); if you want to avoid the possibility of this mistake, make it a normal instance method rather than a local, or if you want to pass in `this` as well, make it a static local function.

File: src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs
Comment (stephentoub) [inline src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs]: This should be made `volatile`.
```suggestion
        private volatile TaskCompletionSource<bool>? _handshakeWaiter;
```

File: src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs
Comment (stephentoub) [inline src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs]: What happens if below we end up not calling ReplyOnReAuthenticationAsync, e.g. because AllowRenegotation is false?  That will leave _handshakeWaiter as non-null, but no one will ever complete it.  Is that going to cause problems?  Can we ensure that if this is set, it'll always be completed/removed?

File: src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs
Comment (stephentoub) [inline src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs]: ConfigureAwait(false)

File: src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs
Comment (stephentoub) [inline src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs]: ConfigureAwait(false)

File: src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs
Comment (wfurt) [reply src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs]: I did not know that, thanks. Would such allocation be visible in VS memory profiler? e.g. is that something I could find eventually?

Comment (stephentoub) [reply src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs]: > Would such allocation be visible in VS memory profiler? 

Yup.

Comment (stephentoub) [reply src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs]: (note the concern here is specifically for async local functions. For other local functions, the compiler can pass referenced state in by ref)

File: src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs
Comment (wfurt) [reply src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs]: I missed that. Since throw, it may not matter but it is cleaner to deal with it explicitly. fixed.

**Comment by wfurt** [comment]
I made an update to change ValuerTask to Task, add missing ConfigureAwait(), named functions as suggested and fix few other issues. I will run another round of ssl-stress and tests in the loop but this should be ready for another round of review in the meantime.

File: src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs
Comment (stephentoub) [inline src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs]: Nit: to avoid indenting all the code above and to make the common case clearer, consider inverting the if condition above such that the break is moved up above and is there only part indented

File: src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs
Comment (stephentoub) [inline src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs]: This is still closing over buffer and outBuffer

File: src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs
Comment (stephentoub) [inline src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs]: A lot of this logic appears to be duplicated with the caller. If that's true, since we expect this to be rare, could we just await a recursive call back to the caller? If there's any concern about stack diving from the recursion, we could queue the call. It'll end up being more expensive when it occurs, but it would seem to simplify things (feel free to tell me I'm wrong :) and this is supposed to be very rare.

File: src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs
Comment (stephentoub) [inline src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs]: In what situation will this be null?

File: src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs
Comment (wfurt) [reply src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs]: updated

File: src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs
Comment (wfurt) [reply src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs]: I updated the logic as you suggested @stephentoub. While working on this I realized we don't have a good way how to test this.  Getting one renegotiation is easy (either with TLS1.3 or with external server) but to do that again is not easy until we have some way how to trigger it in loop with concurrent writes,

File: src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs
Comment (wfurt) [reply src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs]: fixed with recursion rewrite.

File: src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs
Comment (wfurt) [reply src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs]: I took another look and there are only two places when _handshakeWaiter is set to null. This is one of them and second is when renegotiation is disabled - and that is mutually exclusive.

File: src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs
Comment (davidsh) [reply src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs]: I wish we had an API in SslStream (server-side) to trigger a renegotiation. :)  Perhaps it is time to add one.

And while we wait for an API approval, we could perhaps make this API 'internal' and use reflection in our tests to invoke it.

Comment (stephentoub) [reply src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs]: > I wish we had an API in SslStream (server-side) to trigger a renegotiation. :) Perhaps it is time to add one. And while we wait for an API approval, we could perhaps make this API 'internal' and use reflection in our tests to invoke it.

I agree.  However, it seems likely introducing that is going to require some re-work in the locking scheme, since the current scheme in this PR expects that renegotation can only be triggered by reads and relies on that for correctness.  I think we should get this change in, in order to fix the known existing problems and unblock other work, and then introducing a new method (even if internal at first to help with testing) sounds like a good idea.

Comment (wfurt) [reply src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs]: It is on my TODO list. I think I know how to do that at least for schannel and OpenSSL. Also note, that for TLS13 we go through this path even if renegotiation does not exist in that version. When I talked to experts, it seems like that particular status code is used anytime when something needs to be sent to `lsass` for further processing.

**Comment by wfurt** [comment]
It seems like the execution of stress-ssl is broken again as it fails to build. 
cc: @ViktorHofer

**Comment by ViktorHofer** [comment]
cc @Anipik @ericstj

File: src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs
Comment (stephentoub) [inline src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs]:
```suggestion
                        await WriteSingleChunk(writeAdapter, buffer).ConfigureAwait(false);
```

Comment (stephentoub) [reply src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs]: We're still holding on to the rentedBuffer at this point.  Is it rare enough that that's not going to matter?

File: src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs
Comment (stephentoub) [inline src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs]: It'd be helpful to add a comment somewhere detailing the locking scheme being used.  That will, for example, help to explain why it's ok for this nulling out of the field to be done (e.g. because there's only ever one reader at a time, and the reader is responsible for initiating all renegotations, and thus for creating, completing, and removing a waiter).

**Review by stephentoub — APPROVED** [review]

File: src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs
Comment (wfurt) [reply src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs]: updated. I added long comment to where the waiter is created and the logic starts.

File: src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs
Comment (wfurt) [reply src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs]: I was thinking about but nullability give me some grief. I think it would be rare but since you mentioned it, I made update to return the buffer before WriteSingleChunk() is awaited. 
I was also originally looking for way how to loop it back so we do not need to rent new buffer but that goes against the simplicity we discussed.

**Review by stephentoub — APPROVED** [review]

**Comment by stephentoub** [comment]
Thanks, @wfurt!

**Pull request merged**
